crates/uv/src/requirements.rs

```diff
@@ -576,8 +583,160 @@ impl NamedRequirements {
+                        path = path.display(),
+                        name = project.name
+                    );
+                    return Ok(Requirement {
+                        name: project.name,
+                        extras: requirement.extras,
+                        version_or_url: Some(VersionOrUrl::Url(requirement.url)),
+                        marker: requirement.marker,
+                    });
+                }
+
+                // Read Poetry-specific metadata from the `pyproject.toml`.
+                if let Some(tool) = pyproject.tool {
+                    if let Some(poetry) = tool.poetry {
+                        if let Some(name) = poetry.name {
+                            debug!(
+                                "Found Poetry metadata for {path} in `pyproject.toml` ({name})",
+                                path = path.display(),
+                                name = name
+                            );
+                            return Ok(Requirement {
+                                name,
+                                extras: requirement.extras,
+                                version_or_url: Some(VersionOrUrl::Url(requirement.url)),
+                                marker: requirement.marker,
+                            });
+                        }
+                    }
+                }
+            }
+
+            // Attempt to read a `setup.cfg` from the directory.
+            if let Some(setup_cfg) = fs_err::read_to_string(path.join("setup.cfg"))
+                .ok()
+                .and_then(|contents| {
+                    let mut ini = Ini::new_cs();
+                    ini.set_multiline(true);
+                    ini.read(contents).ok()
+                })
+            {
+                if let Some(section) = setup_cfg.get("metadata") {
+                    if let Some(Some(name)) = section.get("name") {
+                        if let Ok(name) = PackageName::from_str(name) {
+                            debug!(
+                                "Found setuptools metadata for {path} in `setup.cfg` ({name})",
+                                path = path.display(),
+                                name = name
+                            );
+                            return Ok(Requirement {
+                                name,
+                                extras: requirement.extras,
+                                version_or_url: Some(VersionOrUrl::Url(requirement.url)),
+                                marker: requirement.marker,
+                            });
+                        }
+                    }
+                }
+            }
+
+            // Attempt to read a `setup.py` from the directory.
+            if let Ok(setup_py) = fs_err::read_to_string(path.join("setup.py")) {
+                static SETUP_PY_NAME: Lazy<Regex> =
+                    Lazy::new(|| Regex::new(r#"name\s*[=:]\s*['"](?P<name>[^'"]+)['"]"#).unwrap());
```

Comment:
\cc @BurntSushi - Here I want to match `setup(name="foo")`-like calls.

Reply:
I do not think frontends are supposed to use build backend–specific heuristics to retrieve metadata.

Reply:
I'm fine with it. These are just heuristics to support non-spec compliant requirement definitions.

Reply:
Why not run `prepare_metadata_for_build_wheel()` and use the metadata from there like pip does?

Reply:
I don't want to pay that cost here when all we need is the project _name_ (which is far less to ask than the project _metadata_). I'll consider it though.

Reply:
The project name is part of the Core metadata spec just like `Requires-Dist` or anything else, but I understand. Thanks for looking into it.

Reply:
Also, it's possible for a build backend to exist that isn't setuptools or poetry and doesn't use PEP 621 metadata.